iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java

```diff
@@ -262,17 +264,52 @@ public void preAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable, E
 
     if (AlterTableType.ADDCOLS.equals(currentAlterTableOp)) {
       Collection<FieldSchema> addedCols =
-          HiveSchemaUtil.schemaDifference(hmsTable.getSd().getCols(), HiveSchemaUtil.convert(icebergTable.schema()));
+          HiveSchemaUtil.newColumns(hmsTable.getSd().getCols(), HiveSchemaUtil.convert(icebergTable.schema()));
+      if (!addedCols.isEmpty()) {
+        updateSchema = icebergTable.updateSchema();
+      }
+      for (FieldSchema addedCol : addedCols) {
+        updateSchema.addColumn(
+            addedCol.getName(),
+            HiveSchemaUtil.convert(TypeInfoUtils.getTypeInfoFromTypeString(addedCol.getType())),
+            addedCol.getComment()
+        );
+      }
+    } else if (AlterTableType.REPLACE_COLUMNS.equals(currentAlterTableOp)) {
+      List<FieldSchema> icebergSchema = HiveSchemaUtil.convert(icebergTable.schema());
+      List<FieldSchema> hmsSchema = hmsTable.getSd().getCols();
+      Collection<FieldSchema> droppedCols = HiveSchemaUtil.newColumns(icebergSchema, hmsSchema);
+      Collection<FieldSchema> addedCols = HiveSchemaUtil.newColumns(hmsSchema, icebergSchema);
+      Collection<FieldSchema> typeUpdated = HiveSchemaUtil.updatedTypeColumns(hmsSchema, icebergSchema);
+      Collection<FieldSchema> commentUpdated = HiveSchemaUtil.updatedCommentColumns(hmsSchema, icebergSchema);
+      if (!droppedCols.isEmpty() || !addedCols.isEmpty() || !typeUpdated.isEmpty() || !commentUpdated.isEmpty()) {
+        updateSchema = icebergTable.updateSchema();
+      }
+
+      for (FieldSchema droppedCol : droppedCols) {
+        updateSchema.deleteColumn(droppedCol.getName());
+      }
+
       for (FieldSchema addedCol : addedCols) {
-        if (updateSchema == null) {
-          updateSchema = icebergTable.updateSchema();
-        }
         updateSchema.addColumn(
             addedCol.getName(),
             HiveSchemaUtil.convert(TypeInfoUtils.getTypeInfoFromTypeString(addedCol.getType())),
             addedCol.getComment()
         );
       }
+
+      for (FieldSchema updatedCol : typeUpdated) {
+        Type newType = HiveSchemaUtil.convert(TypeInfoUtils.getTypeInfoFromTypeString(updatedCol.getType()));
+        if (!(newType instanceof Type.PrimitiveType)) {
```

Comment:
What happens when we try to convert primitive types that are not compatible?

Reply:
The Iceberg library will throw an exception in that case, saying type promotion is not supported (e.g. int -> string)